### DISCUSSION TITLE
TLS support for WebSocket out node

### DISCUSSION THREAD

**Discussion by kitazaki** [phoddie/node-red-mcu#127] [Q&A] [answered]
@phoddie 

@mshioji mentions ssl/tls support on http request node in https://github.com/phoddie/node-red-mcu/issues/120

I think that HTTP and MQTT client nodes support TLS now, but WebSocket out node doesn't support TLS yet.
I was asked the issue from Node-RED user and tried MQTT out node with TLS option, but it displays the error with "unimplemented" message.

(wss://url without TLS option)
![436227944_7687802877993553_2174997001407044876_n](https://github.com/phoddie/node-red-mcu/assets/23477407/61ced63f-fd4f-4d73-b600-9b7d97c5b9f8)

(wss://url with TLS option)
![____________________________2024-05-14_8 46 52_720](https://github.com/phoddie/node-red-mcu/assets/23477407/259914f6-95d2-4117-bdd6-050958b3383a)

Is it possible to support TLS option in WebSocket out node ?
If not, is there any difficulty to support TLS option ?

ws://url without TLS works fine on my environment :-)

![____________________________2024-05-14_8 44 46_480](https://github.com/phoddie/node-red-mcu/assets/23477407/a4be2c3d-2502-49d0-a429-557ed79ad6ef)

![____________________________2024-05-14_8 44 57_720](https://github.com/phoddie/node-red-mcu/assets/23477407/a8dc77c8-bc07-4056-93ad-77fc86db26ca)

**Comment by phoddie** [comment]
@kitazaki – As you observed, TLS is not currently implemented for WebSockets under Node-RED MCU. It is possible. However, it is a little complicated. The challenge is propagating the TLS configuration (certificate, etc) through the stack:

- websocket-client node (which is used by both WebSocket In and WebSocket Out nodes)
- `WebSocket` object (this is a Web-compatible class. It has no way to pass TLS credentials, so we must extend the API)
- ECMA-419 WebSocket client

Both the `WebSocket` object and ECMA-419 WebSocket client have some TLS support. However, they do no support user-specified certificates, which is required for Node-RED MCU Edition.

I have always intended to support TLS for the WebSocket Nodes. If it was easy, it would be done. ;) Maybe I can find some time to take another look.

**Reply by phoddie** [reply]
I had a little time to spend on this today. The situation is not too bad, actually.

- I confirmed that the low-level ECMA-419 websocketclient works with TLS. This just required adding a "wss" configuration, no code changes.
- I confirmed that the WhatWG compatible `WebSocket` also work with TLS when given a `wss:/` URL. Again, this just required adding a "wss" configuration
- Rereading the `WebSocket` implementation I was reminded that its constructor already accepts an options object in place of the `href` argument. That provides a place to pass the TLS configuration through from the Node-RED MCU Edition WebSocket nodes.

So.... this is mostly just a matter of connecting some dots. Fingers crossed. I just need to find some more time to do that....

FWIW – the "chilkat.io" server that the ECMA-419 examples use as an echo server is "ws:" only. It doesn't support "wss:". Someone brought back "[echo.websocket.org/.ws](https://websocket.org/tools/websocket-echo-server/)". It only works with "wss:/". I can use that to test this work.

**Reply by phoddie** [reply]
TLS support for WebSocket In and Out nodes [is committed](https://github.com/phoddie/node-red-mcu/commit/95faaacaa2f840d9979891746a2200b24e31c300). It is only tested with `echo.websocket.org` so please let me know how it goes. (Set up the TLS-Config node in the same way as for secure MQTT connections).

One small change is required in the Moddable SDK for this to work. That isn't live yet, but will be later this week. You can apply the change manually in the meantime. In `$MODDABLE/examples/io/tcp/websocket/WebSocket.js` change [line 60](https://github.com/Moddable-OpenSource/moddable/blob/11b722c599b53f5bfe65237908306e95b6b3af32/examples/io/tcp/websocket/WebSocket.js#L60
) from...

```js
config = device.network.wss;
```
...to:
```js
config = {...(options?.wss ?? device.network.wss)};
```

**Reply by kitazaki** [reply]
@phoddie 
Thank you ! It works fine with my Node-RED WebSocket server !
The url is like this.

![スクリーンショット 2024-05-20 16 03 47](https://github.com/phoddie/node-red-mcu/assets/23477407/666e559a-376d-46c6-be50-96f6cbd67ef3)

It works with another WebSocket service, but the error occurred.
https://piehost.com/ (It's free, but the registration is required.)

The url is like this.
![url_error](https://github.com/phoddie/node-red-mcu/assets/23477407/fa324fb3-7d57-4aea-b555-ae8f4a1eb20c)

msg.payload data is "1716188966040" and it reaches to the piehost by checking its log.

![error](https://github.com/phoddie/node-red-mcu/assets/23477407/0ff087cc-a044-4492-a31f-f2733fb02402)

![piehost](https://github.com/phoddie/node-red-mcu/assets/23477407/b3743be5-f919-430d-8178-3bc04093c5a4)

**Reply by phoddie** [reply]
@kitazaki – Thank you for confirming that the TLS connection is working.

The additional problem you report does not appear to be TLS related. I don't think I can diagnose it without seeing the flow you used. Would you please share that? Thank you.

**Reply by phoddie** [reply]
I am able to reproduce the problem you show **if** there is no "WebSocket In" node for the WebSocket Client **and** it receives a message. I [committed a fix](https://github.com/phoddie/node-red-mcu/commit/c78b8883f8a27b5a9ea87bc52376adc6af342987) for that. Maybe it helps?

**Comment by kitazaki** [comment]
@phoddie 

the flow is below.
I use only "WebSocket out" node with no "WebSocket in" node.

![スクリーンショット 2024-05-21 12 26 58](https://github.com/phoddie/node-red-mcu/assets/23477407/aa675ac5-fd92-4313-8254-ff34b88c4c33)

It works fine for my environment.
I think that it is finished to support TLS for WebSocket nodes.

**Reply by phoddie** [reply, marked as answer]
Thanks for the good news and for sharing the image of the flow. The exception when there is input received but no WebSocket In node is unrelated to TLS. It could happen with an unsecured ws: connection. 

Since this is a Q&A thread, would you please make it as answered?

**Comment by mshioji** [comment]
Hello, @phoddie , @kitazaki ,
Thank you for your great work, Peter-san and Kitazaki-san.

I just curious that if I running WebSocket server on the cloud-Node-RED with "Websocket-In" node by "Listen On", 
I couldn't find TLS settings for it.  *This is not a node-red-mcu issue but  I'm happy if someone knows about this.
( I don't know but It could be a known story? )

I mean...
- Using "WebSocket-Out" with "ws" by "Connect to", I could connect to my cloud-Node-RED "WebSocket-In" like this:
![image](https://github.com/phoddie/node-red-mcu/assets/59914628/f956c54f-7651-4596-a6bb-047dcb13570b)

- then switch it to "wss" , TLS settings are displayed as you know.  
![image](https://github.com/phoddie/node-red-mcu/assets/59914628/c755ee37-29d7-4a5f-af1d-ba795d94c868)

- however, on the server side, there are nothing about TLS settings like this, it seems no place to set key and cert. 
![image](https://github.com/phoddie/node-red-mcu/assets/59914628/9d3e7d66-6920-498f-b114-96699c36ed47)
![image](https://github.com/phoddie/node-red-mcu/assets/59914628/92c51cb4-15b3-402e-b912-9e961917222f)

(In a similar case, aedes MQTT Broker node has an explicit TLS setting.)
![image](https://github.com/phoddie/node-red-mcu/assets/59914628/6285d9fd-7727-49a1-852d-d3e023d848fc)

Thank you, have a good day!

**Reply by phoddie** [reply]
It appears that the Node-RED editor does not allow a TLS configuration for a WebSocket listener. Even if it did, the WebSocket and HTTP servers ("listeners") in Node-RED MCU Edition do not currently support TLS.

**Reply by mshioji** [reply]
Hi @phoddie , Thanks for make it clear.
I understand it seems we should look for another solution from Mode-RED if we would like to setup server with wss.
I think this is a server side issue, therefore for MCU, I believe WebSocket Out support is enough. 

Thank you very much,

**Reply by kitazaki** [reply]
@mshioji 
WebSocket listener is using same port as Node-RED.
Node-RED configuration is defined in settings.js as below.

https: {
  key: require("fs").readFileSync('/data/certs/privkey.pem'),
  cert: require("fs").readFileSync('/data/certs/fullchain.pem')
},

Secure "WebSocket in" node depends on that TLS setting.

"MQTT Broker" node is using different port from Node-RED.
1883 is for MQTT and 8883 is for MQTTS as default, so it has TLS setting its own.

**Reply by kitazaki** [reply]
@mshioji 

add https configuration to settings.js in ~/.node-red

`
https: {
  key: require("fs").readFileSync('/data/certs/privkey.pem'),
  cert: require("fs").readFileSync('/data/certs/fullchain.pem')
},
`

run Node-RED (below is docker case)

`
$ sudo docker run -d -p 443:1880 -v /home/ubuntu/.node-red:/data nodered/node-red:latest
`

![スクリーンショット 2024-05-22 8 44 28](https://github.com/phoddie/node-red-mcu/assets/23477407/078a301f-b85a-43fb-8952-ee4dbbcfb103)

**Comment by mshioji** [comment]
Hi @kitazaki -san,

Thank you for the information, oh, Node-RED settings.js works for it!
I will try it later in my environment.

... there are many options in settings.js! ^^
Thank you!
